src/gui/tray/Window.qml

```diff
@@ -10,6 +10,125 @@ import QtGraphicalEffects 1.0
+            }
+        } else {
+            tbOrientation = systrayBackend.taskbarOrientation();
+        }
+        if (Qt.platform.os === "osx") {
+            taskbarHeight = 22;
+            taskbarWidth = Screen.width;
+        } else if (Qt.platform.os === "linux") {
+            taskbarHeight = (tbOrientation === 0 || tbOrientation === 2) ? 32 : Screen.height;
+            taskbarWidth = (tbOrientation === 0 || tbOrientation === 2) ? Screen.width : 32;
+        } else {
+            taskbarHeight = systrayBackend.taskbarRect().height;
+            taskbarWidth = systrayBackend.taskbarRect().width;
+        }
+
+        switch(tbOrientation) {
+            // Platform separation here: Windows and macOS draw coordinates have to be given in screen-coordinates
+            // KDE and most xorg based DEs expect them as virtual coordinates
+            case 0:
+                console.debug("Taskbar is on the bottom.");
+                trayWindowX = trayIconCenter.x - trayWindow.width / 2;
+                trayWindowY = (Qt.platform.os !== "linux") ? (Screen.height - taskbarHeight - trayWindow.height - 4)
+                                                           : (Screen.height + Screen.virtualY - taskbarHeight - trayWindow.height - 4);
+                break;
+            case 1:
+                console.debug("Taskbar is on the left.");
+                trayWindowX = (Qt.platform.os !== "linux") ? (taskbarWidth + 4)
+                                                           : (Screen.virtualX + taskbarWidth + 4);
+                trayWindowY = trayIconCenter.y;
+                break;
+            case 2:
+                console.debug("Taskbar is on the top.");
+                trayWindowX = trayIconCenter.x - trayWindow.width / 2;
+                trayWindowY = Screen.virtualY + taskbarHeight + 4;
+                break;
+            case 3:
+                console.debug("Taskbar is on the right.");
+                trayWindowX = (Qt.platform.os !== "linux") ? (Screen.width - taskbarWidth - trayWindow.width - 4)
+                                                           : (Screen.width + Screen.virtualX - taskbarWidth - trayWindow.width - 4);
+                trayWindowY = trayIconCenter.y;
+                break;
+        }
+
+        console.debug("Screen.height:",Screen.height);
+        console.debug("Screen.desktopAvailableHeight:",Screen.desktopAvailableHeight);
+        console.debug("Screen.virtualY:",Screen.virtualY);
+        console.debug("Screen.width:",Screen.width);
+        console.debug("Screen.desktopAvailableWidth:",Screen.desktopAvailableWidth);
+        console.debug("Screen.virtualX:",Screen.virtualX);
+        console.debug("Taskbar height:",taskbarHeight);
+        console.debug("Taskbar width:",taskbarWidth);
+
+        if ( (Screen.width <= trayWindowX + trayWindow.width)) {
+            console.debug("Out-of-screen condition on the right detected. Adjusting window position.");

```

Comment:
It probably could, but I rather decided to have the whole positioning logic in one place and keep it readable. Do you disagree with this? Then I'd need some suggestions b/c I don't see it that way.

Reply:
In case I was unclear, I mean that it feels like you do that twice. Once before the bunch of console.debug calls (ideal positioning) and once after ("oops that turned out wrong let's change our mind"). I wonder if those two couldn't be merged.

Especially it feels like this code is written slightly too low level and that might be why it turns out like this. This kind of job I generally end up pulling something like Qt.vector2d to compute the points in term of vector operations. I'm pretty sure it'd improve things a lot here... what you probably want is to pick a corner of the window and position it relative to the icon center + coordinate transform on linux... but I got to admit it's too late in the night for me to draft it.

Reply:
I get your point, but am still unsure about your preferred alternatives. IMHO if it was just positioning with regard to the tray icon this would be still readable when consolidated, but we have to deal with positioning relative to the taskbar plus hardy platform-dependent taskbar layout (e.g. the taskbar can have different widths, but the icon is not necessarily centered - multi-row icons in wider taskbars as in KDE and Windows) as well as screen border collision checks that are platform dependent too because most Linux DEs (and/or: the Qt framework layer on top of it) handle screen coordinates with regard to virtual and absolute positions differently. You could still build an own logic-layer in the backend that prevents all those quirks by expanding/correcting the Qt-API results, so that you only pull in one pre-calculated screen geometry in QML. 

I adressed all the other points, which reduced some code in QML and transferred it's logic to C++, but leave the out-of-screen check as it is, leaving the design up to you.

Reply:
Sure thing. I think I need to at least draft something on that one but didn't have the energy last night, will try to do it today.

Thanks a lot for the other adjustments! I'll get to properly review them a bit later today as well.

Reply:
So I'm working on a couple of commits, I'll push them in the branch once they are ready (they'll likely require further testing though, I don't have enough gear locally to test as I'd like). I hope you don't mind.